pyout/common.py

```diff
@@ -0,0 +1,395 @@
+"""
+
+from collections import defaultdict, Mapping, Sequence
+from functools import partial
+import inspect
+
+from pyout import elements
+from pyout.field import Field, Nothing
+
+NOTHING = Nothing()
+
+
+class RowNormalizer(object):
+    """Transform various input data forms to a common form.
+
+    An un-normalized can be one of three kinds:
+
+      * a mapping from column names to keys
+
+      * a sequence of values in the same order as `columns`
+
+      * any other value will be taken as an object where the column values can
+        be accessed via an attribute with the same name
+
+    To normalized a row, it is
+
+      * converted to a dict that maps from column names to values
+
+      * all callables are stripped out and replaced with their initial values
+
+      * if the value for a column is missing, it is replaced with a Nothing
+        instance whose value is specified by the column's style (an empty
+        string by default)
+
+    Parameters
+    ----------
+    columns : sequence of str
+        Column names.
+    style : dict, optional
+        Column styles.
+
+    Attributes
+    ----------
+    methods : callable
+        A function that takes a row and returns a normalized one.  This is
+        chosen at time of the first call.  All subsequent calls should use the
+        same kind of row.
+    nothings : dict
+        Maps column name to the placeholder value to use if that column is
+        missing.
+    """
+
+    def __init__(self, columns, style):
+        self._columns = columns
+        self.method = None
+
+        self.delayed = defaultdict(list)
+        self.delayed_columns = set()
+        self.nothings = {}  # column => missing value
+
+        for column in columns:
+            cstyle = style[column]
+
+            if "delayed" in cstyle:
+                value = cstyle["delayed"]
+                group = column if value is True else value
+                self.delayed[group].append(column)
+                self.delayed_columns.add(column)
+
+            if "missing" in cstyle:
+                self.nothings[column] = Nothing(cstyle["missing"])
+            else:
+                self.nothings[column] = NOTHING
+
+
+    def __call__(self, row):
+        """Normalize `row`
+
+        Parameters
+        ----------
+        row : mapping, sequence, or other
+            Data to normalize.
+
+        Returns
+        -------
+        A tuple (callables, row), where `callables` is a list (as returned by
+        `strip_callables`) and `row` is the normalized row.
+        """
+        if self.method is None:
+            self.method = self._choose_normalizer(row)
+        return self.method(row)
+
+    def _choose_normalizer(self, row):
+        if isinstance(row, Mapping):
+            return partial(self._normalize, self.getter_dict)
+        if isinstance(row, Sequence):
+            return partial(self._normalize, self.getter_seq)
+        return partial(self._normalize, self.getter_attrs)
```

Comment:
I typically go for a solution with smaller visual clutter due to repeating pattern (patterns of the same code disturb my visual system) and thus avoiding code duplication:
```python
if isinstance(row, Mapping):
    getter = self.getter_dict
elif isinstance(row, Sequence):
    getter = self.getter_seq
else:
    getter = self.getter_attrs
return partial(self._normalize, getter)
```
allows me to avoid analysis of the code if there is anything else different but figuring out the getter to use.  Even this one has too much repeating pattern... so I could have done:

```python
for cls, getter in [
   (Mapping, self.getter_dict),
   (Sequence, self.getter_seq),
   (object, self.getter_attrs)
]:
   if isinstance(row, cls):
         return partial(self._normalize, getter)
   raise ValueError("Works only with subclasses of object or dict/list like interfaces. Got %r" % row)
```

Reply:
actually seems everything is instance of object:
```ipython
In [1]: class A:
   ...:      pass
   ...: 

In [2]: a = A

In [3]: isinstance(A, object)
Out[3]: True
```
now so last exception could be just removed or RuntimeError("Wrong assumptions ;)")

Reply:
> I typically go for a solution with smaller visual clutter due to repeating pattern (patterns of the same code disturb my visual system) and thus avoiding code duplication:

As do I.  Thanks for spotting that I failed to do that here.  I will go with your first suggestion.  IMO your second suggestion compromises readability for the sake of non-redundancy.